rust/main/agents/relayer/src/msg/metadata/message_builder.rs

```diff
@@ -0,0 +1,430 @@
+            MessageMetadataBuilder::new(Arc::new(Box::new(base_builder)), ism_address, &message)
+                .await
+                .expect("Failed to build MessageMetadataBuilder");
+
+        message_builder.depth = ISM_MAX_DEPTH;
+
+        let res = build_message_metadata(message_builder, ism_address, &message)
+            .await
+            .expect("Metadata building failed");
+
+        match res.metadata {
+            Metadata::Found(_) => {
+                panic!("Metadata found when it should have failed");
+            }
+            Metadata::Failed(err) => {
+                assert_eq!(err, MetadataBuildError::MaxIsmDepthExceeded(ISM_MAX_DEPTH));
+            }
+        }
+    }
+
+    #[tokio::test]
+    async fn ism_count_already_reached() {
+        let base_builder = build_mock_base_builder();
+        insert_mock_routing_ism(&base_builder);
+
+        let ism_address = H256::zero();
+        let message = HyperlaneMessage::default();
+
+        let message_builder =
+            MessageMetadataBuilder::new(Arc::new(Box::new(base_builder)), ism_address, &message)
+                .await
+                .expect("Failed to build MessageMetadataBuilder");
+
+        *message_builder.ism_count.lock().await = ISM_MAX_COUNT;
+
+        let res = build_message_metadata(message_builder, ism_address, &message)
+            .await
+            .expect("Metadata building failed");
+
+        match res.metadata {
+            Metadata::Found(_) => {
+                panic!("Metadata found when it should have failed");
+            }
+            Metadata::Failed(err) => {
+                assert_eq!(err, MetadataBuildError::MaxIsmCountReached(ISM_MAX_COUNT));
+            }
+        }
+    }
+
+    #[tokio::test]
+    async fn max_depth_reached() {
+        let base_builder = build_mock_base_builder();
+
+        for _ in 0..3 {
+            insert_mock_routing_ism(&base_builder);
+        }
+
+        let ism_address = H256::zero();
+        let message = HyperlaneMessage::default();
+
+        let mut message_builder =
+            MessageMetadataBuilder::new(Arc::new(Box::new(base_builder)), ism_address, &message)
+                .await
+                .expect("Failed to build MessageMetadataBuilder");
+        message_builder.depth = ISM_MAX_DEPTH - 2;
+
+        let res = build_message_metadata(message_builder, ism_address, &message)
+            .await
+            .expect("Metadata building failed");
+
+        match res.metadata {
+            Metadata::Found(_) => {
+                panic!("Metadata found when it should have failed");
+            }
+            Metadata::Failed(err) => {
+                assert_eq!(err, MetadataBuildError::MaxIsmDepthExceeded(ISM_MAX_DEPTH));
+            }
+        }
+    }
+
+    #[tokio::test]
+    async fn max_ism_count_reached() {
+        let base_builder = build_mock_base_builder();
+
+        for _ in 0..11 {
+            insert_mock_routing_ism(&base_builder);
+        }
```

Comment:
I think using the routing ISM here can make it less easy to understand because depth will always be equal to the ism count. An aggregation ISM that points to a bunch of other ISMs would probably be more common and easier to understand. In the unit test I'd add a check that depth is 1 and count is the number of sub-isms + 1, so the reader has an easier time figuring out what the test does